hail/python/hailtop/batch/batch_pool_executor.py

```diff
@@ -537,9 +546,10 @@ async def async_result(self, timeout: Optional[Union[float, int]] = None):
             raise concurrent.futures.CancelledError()
         try:
             return await asyncio.wait_for(self.fetch_coro, timeout=timeout)
-        except asyncio.TimeoutError:
-            await self.async_cancel()
-            raise
+        except asyncio.TimeoutError as e:
+            raise concurrent.futures.TimeoutError() from e
+        finally:
+            await self.batch.cancel()
```

Comment:
This only fires if someone calls async_result, right? I'm sorry I was a bit vague. When I said `fetch_coro`, I was thinking of `self._async_fetch_result`, which is the function called to start the fetch_coro. I think you add self.batch.cancel() to its finally block, then if fetch_coro is cancelled, the batch will be cancelled. The batch will also be cancelled if any error happens during result fetching or job waiting.